### PR TITLE
Sap hana review

### PR DESCRIPTION
**Pull request by siddharth7**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 1]
```diff
@@ -57,50 +59,87 @@ def load_arguments(self, _):
...
+
+    with self.argument_context('backup protectable-item') as c:
+        c.argument('container_name', container_name_type)
+
     # Policy
     with self.argument_context('backup policy') as c:
         c.argument('vault_name', vault_name_type, id_part='name')
 
-    for command in ['show', 'delete', 'list-associated-items']:
+    for command in ['show', 'delete', 'list-associated-items', 'set', 'new']:
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:108]: why were set and new added here?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 2]
```diff
@@ -57,50 +59,87 @@ def load_arguments(self, _):
...
         with self.argument_context('backup policy ' + command) as c:
             c.argument('name', policy_name_type, options_list=['--name', '-n'], help='Name of the backup policy. You can use the backup policy list command to get the name of a policy.')
 
     with self.argument_context('backup policy set') as c:
         c.argument('policy', type=file_type, help='JSON encoded policy definition. Use the show command with JSON output to obtain a policy object. Modify the values using a file editor and pass the object.', completer=FilesCompleter())
 
+    with self.argument_context('backup policy new') as c:
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:115]: how were new policies made earlier?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 3]
```diff
@@ -57,50 +59,87 @@ def load_arguments(self, _):
...
+        c.argument('policy', type=file_type, help='JSON encoded policy definition. Use the show command with JSON output to obtain a policy object. Modify the values using a file editor and pass the object.', completer=FilesCompleter())
+        c.argument('workload_type', wl_type)
+        c.argument('container_type', backup_management_type)
+
     with self.argument_context('backup policy list') as c:
-        c.argument('vault_name', vault_name_type, id_part=None)
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:88]: why is vault_name removed?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 4]
```diff
@@ -115,14 +154,26 @@ def load_arguments(self, _):
             c.argument('item_name', item_name_type)
 
     with self.argument_context('backup protection backup-now') as c:
-        c.argument('retain_until', type=datetime_type, help='The date until which this backed up copy will be available for retrieval, in UTC (d-m-Y).')
+        c.argument('retain_until', type=datetime_type, help='The date until which this backed up copy will be available for retrieval, in UTC (d-m-Y).', options_list=['--retain-until', '-rt'])
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:157]: is -rt option from PM spec?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 5]
```diff
@@ -115,14 +154,26 @@ def load_arguments(self, _):
...
+        c.argument('backup_type', arg_type=get_enum_type(['Full', 'Differential', 'Log', 'CopyOnlyFull']), options_list=['--backup-type', '-bt'])
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py:158]: similarly, other shorthands?

File: src/azure-cli/azure/cli/command_modules/backup/_params.py [hunk 6]
```diff
@@ -153,3 +207,14 @@ def load_arguments(self, _):
 
     with self.argument_context('backup job wait') as c:
         c.argument('timeout', type=int, help='Maximum time, in seconds, to wait before aborting.')
+
+    with self.argument_context('backup recoveryconfig') as c:
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/_params.py]: why are the parameters distributed in 2 groups?

File: src/azure-cli/azure/cli/command_modules/backup/_validators.py [hunk 7]
```diff
@@ -10,10 +10,10 @@
 
 def datetime_type(string):
     """ Validates UTC datettime in accepted format. Examples: 31-12-2017 """
-    accepted_date_formats = ['%d-%m-%Y']
+    accepted_date_formats = ['%d-%m-%Y', '%d-%m-%Y-%H:%M:%S']
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/_validators.py:13]: keep the 2 error messages separate. Given a scenario, the user might just want to see one format as the correct format

File: src/azure-cli/azure/cli/command_modules/backup/commands.py [hunk 8]
```diff
@@ -7,72 +7,90 @@
...
-    with self.command_group('backup policy', backup_custom, client_factory=protection_policies_cf) as g:
+    with self.command_group('backup policy', backup_custom_base, client_factory=protection_policies_cf) as g:
         g.command('get-default-for-vm', 'get_default_policy_for_vm')
-        g.show_command('show', 'show_policy')
-        g.command('list', 'list_policies', client_factory=backup_policies_cf, table_transformer=transform_policy_list)
+        g.show_command('show', 'show_policy', table_transformer=transform_wl_policy_show)
+        g.command('list', 'list_policies', client_factory=backup_policies_cf, table_transformer=transform_policies_list)
         g.command('list-associated-items', 'list_associated_items_for_policy', client_factory=backup_protected_items_cf, table_transformer=transform_item_list)
-        g.command('set', 'set_policy')
+        g.command('set', 'set_policy', table_transformer=transform_wl_policy_set)
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/commands.py:53]: do we need a table transformer for set policy?

File: src/azure-cli/azure/cli/command_modules/backup/custom.py [hunk 9]
```diff
@@ -497,6 +512,14 @@ def _is_native_name(name):
     return ";" in name
 
 
+def _is_range_valid(start_date, end_date):
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/custom.py]: remove this function

Comment (sahajb) [reply src/azure-cli/azure/cli/command_modules/backup/custom.py]: Done

File: src/azure-cli/azure/cli/command_modules/backup/commands.py (hunk 8 above)
Comment (sahajb) [reply src/azure-cli/azure/cli/command_modules/backup/commands.py:53]: Yes given in spec

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 4 above)
Comment (sahajb) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:157]: No added by me. Will ask PM to confirm all shorthands once

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 5 above)
Comment (sahajb) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py:158]: Most of them are. For those that are not will confirm.

File: src/azure-cli/azure/cli/command_modules/backup/_params.py (hunk 6 above)
Comment (sahajb) [reply src/azure-cli/azure/cli/command_modules/backup/_params.py]: Optional ones are in second group. Can be merged as well provided no other sub command for recoveryconfig is needed.

File: src/azure-cli/azure/cli/command_modules/backup/_validators.py (hunk 7 above)
Comment (sahajb) [reply src/azure-cli/azure/cli/command_modules/backup/_validators.py:13]: Need to discuss

File: src/azure-cli/azure/cli/command_modules/backup/custom_help.py [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+    paged_containers = client.list(vault_name, resource_group_name, filter_string)
+    containers = _get_list_from_paged_response(paged_containers)
+
+    if container_name and _is_native_name(container_name):
+        return [container for container in containers if container.name == container_name]
+
+    return containers
+
+
+def _get_protectable_item_for_vm(cli_ctx, vault_name, vault_rg, vm_name, vm_rg):
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/custom_help.py]: remove the vm specific functions from this file

File: src/azure-cli/azure/cli/command_modules/backup/tests/latest/preparers.py [hunk 11]
```diff
@@ -89,9 +89,10 @@ def create_resource(self, name, **kwargs):
         if not self.dev_setting_value:
             self.resource_group = self._get_resource_group(**kwargs)
             self.location = self._get_resource_group_location(**kwargs)
-            param_format = '-n {} -g {} --image {} --admin-username {} --admin-password {}'
+            param_format = '-n {} -g {} --image {} --admin-username {} --admin-password {} --tags {}'
             param_string = param_format.format(name, self.resource_group, 'Win2012R2Datacenter', name,
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/tests/latest/preparers.py:93]: Update it to param_string = param_format.format(name, self.resource_group, 'Win2012R2Datacenter', name,
                                               '%j^VYw9Q3Z@Cu$*h', 'MabUsed=Yes Owner=sisi Purpose=CLITest DeleteBy=12-2099 AutoShutdown=No')

Comment (sahajb) [reply src/azure-cli/azure/cli/command_modules/backup/tests/latest/preparers.py:93]: Done

File: src/azure-cli/azure/cli/command_modules/backup/custom_help.py (hunk 10 above)
Comment (sahajb) [reply src/azure-cli/azure/cli/command_modules/backup/custom_help.py]: Done

File: src/azure-cli/azure/cli/command_modules/backup/custom_wl.py [hunk 12]
```diff
@@ -0,0 +1,710 @@
...
+            Please pass a native container name.
+            """)
+
+    # Trigger unregister and wait for completion
+    result = sdk_no_wait(True, client.unregister,
+                         vault_name, resource_group_name, fabric_name, container_name)
+    return cust_help._track_register_operation(cmd.cli_ctx, result, vault_name, resource_group_name, container_name)
+
+
+def show_wl_container(cmd, name, resource_group_name, vault_name, container_type, status="Registered"):
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:161]: can we add a check for native name here?

File: src/azure-cli/azure/cli/command_modules/backup/custom_wl.py [hunk 13]
```diff
@@ -0,0 +1,710 @@
...
+    return protection_containers_cf(cmd.cli_ctx).get(vault_name, resource_group_name, fabric_name, name)
+
+
+def list_wl_containers(client, resource_group_name, vault_name, container_type, status="Registered"):
+    return cust_help._get_containers(client, container_type, status, resource_group_name, vault_name)
+
+
+def show_wl_item(cmd, resource_group_name, vault_name, container_name, name):
```
Comment (siddharth7) [inline src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:176]: can we add a check for native name here?

File: src/azure-cli/azure/cli/command_modules/backup/custom_wl.py (hunk 12 above)
Comment (sahajb) [reply src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:161]: Done

File: src/azure-cli/azure/cli/command_modules/backup/custom_wl.py (hunk 13 above)
Comment (sahajb) [reply src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:176]: Done

**Pull request closed without merging**